### PR TITLE
chap 7 with basic signup

### PR DESCRIPTION
**Pull request by minhngoc826**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,38 +1,37 @@
...
-gem 'bcrypt',         '3.1.11'
-gem 'bootstrap-sass', '3.3.6'
-gem 'puma',         '3.4.0'
-gem 'sass-rails',   '5.0.6'
-gem 'uglifier',     '3.0.0'
-gem 'coffee-rails', '4.2.1'
-gem 'jquery-rails', '4.1.1'
-gem 'turbolinks',   '5.0.1'
-gem 'jbuilder',     '2.4.1'
+gem "rails",        "5.0.1"
```
Comment (hungnh103) [inline Gemfile]: remove whitespace
please refer: https://github.com/framgia/coding-standards/blob/master/vn/README.md#ruby-on-rails

File: app/assets/javascripts/static_pages.coffee [hunk 2]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
```
Comment (hungnh103) [inline app/assets/javascripts/static_pages.coffee:1]: please remove this file by using git command:
`git rm <filename>`

File: app/assets/javascripts/users.coffee [hunk 3]
```diff
@@ -1,3 +0,0 @@
-# Place all the behaviors and hooks related to the matching controller here.
```
Comment (hungnh103) [inline app/assets/javascripts/users.coffee:1]: as above mentioned

File: app/assets/stylesheets/custom.scss [hunk 4]
```diff
@@ -1,9 +1,12 @@
 @import "bootstrap-sprockets";
 @import "bootstrap";
 
+$gray-medium-light: #eaeaea;
+$state-danger-text: red;
+
```
Comment (hungnh103) [inline app/assets/stylesheets/custom.scss]: remove this line

File: app/assets/stylesheets/custom.scss [hunk 5]
```diff
@@ -56,7 +62,7 @@ p {
 	padding-top: 9px;
 	font-weight: bold;
 	&:hover {
-		color: #fff;
+		color: yellow;
```
Comment (hungnh103) [inline app/assets/stylesheets/custom.scss]: indent

File: app/assets/stylesheets/custom.scss [hunk 6]
```diff
@@ -89,4 +95,90 @@ footer {
 
 img {
 	display: none;
-}
\ No newline at end of file
+}
+
+
+@mixin box_sizing {
```
Comment (hungnh103) [inline app/assets/stylesheets/custom.scss]: remove this line

File: app/assets/stylesheets/custom.scss [hunk 7]
```diff
@@ -89,4 +95,90 @@ footer {
...
+  -moz-box-sizing:    border-box;
```
Comment (hungnh103) [inline app/assets/stylesheets/custom.scss]: remove whitespace

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -1,6 +1,9 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  include UsersHelper
+
   def hello
```
Comment (hungnh103) [inline app/controllers/application_controller.rb]: action này anh có dùng nữa ko ạ, nếu ko thì xóa đi anh nhé

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -1,4 +1,31 @@
 class UsersController < ApplicationController
   def new
+    @user = User.new
   end
+
+  def show
+    @user = User.find_by id: params[:id]
```
Comment (hungnh103) [inline app/controllers/users_controller.rb:7]: use call back for this block code
please refer more at:
https://viblo.asia/LeAnhTuan/posts/MLzkOYQRGpq
http://guides.rubyonrails.org/active_record_callbacks.html

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -1,4 +1,31 @@
...
+    if @user.nil?
+      render file: "public/404.html", layout: false
+    end
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      flash[:success] = t "controllers.users.wellcome"
```
Comment (hungnh103) [inline app/controllers/users_controller.rb]: anh viết I18n ngắn gọn như này thôi nhé: `t ".welcome"`

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -1,4 +1,31 @@
...
+      log_in @user
+      redirect_to @user
+    else
+      flash[:danger] = t "controllers.users.fail"
```
Comment (hungnh103) [inline app/controllers/users_controller.rb]: as above mentioned

File: app/helpers/application_helper.rb [hunk 12]
```diff
@@ -1,17 +1,6 @@
...
-  #     base_title
-  #   else
-  #     page_title + " | " + base_title
-  #   end
-  # end
-
-  def full_title(page_title = '')
-    base_title = "Ruby on Rails Tutorial Sample App"
+  def full_title page_title = ""
+    base_title = t "helpers.app.ror"
```
Comment (hungnh103) [inline app/helpers/application_helper.rb]: as above mentioned

File: app/helpers/users_helper.rb [hunk 13]
```diff
@@ -1,2 +1,7 @@
 module UsersHelper
+  def gravatar_for user, size: 80
```
Comment (hungnh103) [inline app/helpers/users_helper.rb:2]: please use gem `config`
refer more at: https://github.com/railsconfig/config

File: app/models/user.rb [hunk 14]
```diff
@@ -1,12 +1,13 @@
...
-	VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
-  	validates :email, presence: true, length: { maximum: 50},
-                    format: { with: VALID_EMAIL_REGEX },
-                    # uniqueness: true
-                    uniqueness: { case_sensitive: false }
-    has_secure_password
-    validates :password, length: { minimum: 8, maximum: 20 }, presence: true
+  before_save {email.downcase!}
+
+  validates :name, presence: true, length: {maximum: 30}
```
Comment (hungnh103) [inline app/models/user.rb]: use gem `config`

File: app/views/layouts/_footer.html.erb [hunk 15]
```diff
@@ -1,13 +1,16 @@
...
+    </a>
+    by <a href="http://www.michaelhartl.com/"><%= t "footer.author" %></a>
   </small>
   <nav>
     <ul>
-      <li><%= link_to "About",   about_path %></li>
-      <li><%= link_to "Contact", contact_path %></li>
+      <li><%= link_to t("header.menu.about"), about_path %></li>
+      <li><%= link_to t("footer.contact.title"), contact_path %></li>
       <li><a href="http://news.railstutorial.org/">News</a></li>
```
Comment (hungnh103) [inline app/views/layouts/_footer.html.erb]: I18n

File: app/views/layouts/_header.html.erb [hunk 16]
```diff
@@ -1,13 +1,10 @@
...
   <div class="container">
-    <%= link_to "sample app", '#', id: "logo", class: "col-md-5 col-sm-5 col-xs-4" %>
-    <nav>
-      <ul class="nav navbar-nav navbar-right col-md-7">
-        <li><%= link_to "Home",   root_path %></li>
-        <li><%= link_to "Help",   help_path %></li>
-        <li><%= link_to "Log in", '#' %></li>
-      </ul>
-    </nav>
+    <%= link_to t("header.logo"), root_path, id: "logo"%>
```
Comment (hungnh103) [inline app/views/layouts/_header.html.erb:3]: missing whitespace

File: app/views/layouts/application.html.erb [hunk 17]
```diff
@@ -1,20 +1,28 @@
...
-  <body>
-  		<%= render 'layouts/header' %>
-  		<div class="container col-md-12">
-  			<%= yield %>
-  		</div>
-  		<%= render 'layouts/footer' %>
-    
-  </body>
+  <div class="container">
+    <!-- #show flass -->
```
Comment (hungnh103) [inline app/views/layouts/application.html.erb]: remove comment

File: app/views/layouts/application.html.erb [hunk 18]
```diff
@@ -1,20 +1,28 @@
...
+      <%= content_tag :div, message, class: "alert alert-#{_}" %>
+    <% end %>
+
+    <%= yield %>
+
+    <%= render "layouts/footer" %>
+
+    <%= debug(params) if Rails.env.development? %>
+  </div>
+</body>
```
Comment (hungnh103) [inline app/views/layouts/application.html.erb]: indent

File: app/views/users/new.html.erb [hunk 19]
```diff
@@ -1,2 +1,22 @@
-<h1>Users#new</h1>
-<p>Find me in app/views/users/new.html.erb</p>
+<%= provide :title, t("users.new.signup") %>
```
Comment (hungnh103) [inline app/views/users/new.html.erb]: remove `=`

File: config/locales/en.yml [hunk 20]
```diff
@@ -1,23 +1,76 @@
...
+      sample: " sample application."
+      signup: "Sign up now!"
+      logo: "Rails logo"
+
+  users:
+    new:
+      signup: "Sign up"
+      confirm: "Confirmation"
+      create: "Create your account"
+
```
Comment (hungnh103) [inline config/locales/en.yml]: remove lines

File: app/assets/stylesheets/custom.scss (hunk 4 above)
Comment (minhngoc826) [reply app/assets/stylesheets/custom.scss]: xóa dòng nào e? dòng trắng ah

**Pull request merged**